### PR TITLE
Make os:cmd calls stricter by not ignoring exit status

### PR DESCRIPTION
**Pull request by lucafavatellawork**
Preparation for #38

### CODE REVIEW

**Comment by lucafavatellawork** [comment]
I updated #38 , using that #38 I built Docker image, using https://github.com/opensparkl/sse_cli/pull/23 I ran sse_cli tests on Docker image, and they passed ("53 passed in 12.52s").

**Comment by lucafavatellawork** [comment]
> I updated #38 , using that #38 I built Docker image, using [opensparkl/sse_cli#23](https://github.com/opensparkl/sse_cli/pull/23) I ran sse_cli tests on Docker image, and they passed ("53 passed in 12.52s").

I ran again those tests, commenting the Docker stop&rm. I then ran `more yaws/report.log` in the container, I eye-balled it, and seemed ok. I then did `docker logs ...` and I saw no error - notice the status logged:
```

=INFO REPORT==== 17-Dec-2019::11:05:59 ===
    module: sse_util
    line: 712
    pid: <0.34.0>
    cmd: "mkdir -p ssl"
    status: 0
    result: []

=INFO REPORT==== 17-Dec-2019::11:05:59 ===
    module: sse_util
    line: 712
    pid: <0.34.0>
    cmd: "sh /usr/local/lib/sse/lib/sse_yaws-1.3.4-43-g71af/priv/openssl/gen_server_certs.sh   /usr/local/lib/sse/lib/sse_yaws-1.3.4-43-g71af/priv/openssl/openssl.conf ssl"
    status: 0
    result: "Generating RSA private key, 2048 bit long modulus\n...................+++++\n...+++++\ne is 65537 (0x010001)\nUsing configuration from openssl.conf\nCheck that the request matches the signature\nSignature ok\nThe Subject's Distinguished Name is as follows\ncommonName            :ASN.1 12:'4bc87d162bc0'\nCertificate is to be certified until Dec 16 11:05:59 2020 GMT (365 days)\n\nWrite out database with 1 new entries\nData Base Updated\nGenerating RSA private key, 2048 bit long modulus\n..................................................................................+++++\n................+++++\ne is 65537 (0x010001)\nUsing configuration from openssl.conf\nCheck that the request matches the signature\nSignature ok\nThe Subject's Distinguished Name is as follows\ncommonName            :ASN.1 12:'4bc87d162bc0'\norganizationName      :ASN.1 12:'sse'\nCertificate is to be certified until Dec 16 11:05:59 2020 GMT (365 days)\n\nWrite out database with 1 new entries\nData Base Updated\n"

=INFO REPORT==== 17-Dec-2019::11:05:59 ===
    module: sse_util
    line: 712
    pid: <0.34.0>
    cmd: "mkdir -p log/otp"
    status: 0
    result: []

=INFO REPORT==== 17-Dec-2019::11:05:59 ===
    module: sse_util
    line: 712
    pid: <0.34.0>
    cmd: "mkdir -p log/yaws"
    status: 0
    result: []
...
```

**Comment by lucafavatellawork** [comment]
Assigning to me while updating `sse_util` as per docs comments by @jacobythwaites

File: apps/sse/src/sse_util_os.erl [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+        case maps:get(exit_status, Opts, false) of
+            false -> false;
+            _ -> true
+        end,
+    {SpawnCmd, SpawnOpts, SpawnInput, Eot} = mk_cmd(os:type(), ExitStatusOpt, validate(Cmd)),
+    Port = open_port({spawn, SpawnCmd}, [binary, stderr_to_stdout,
+                                         stream, in, hide | SpawnOpts]),
+    MonRef = erlang:monitor(port, Port),
+    true = port_command(Port, SpawnInput),
+    {Status, Bytes} = get_data(Port, MonRef, Eot, ExitStatusOpt, [], 0, maps:get(max_size, Opts, infinity)),
```
Comment (jacobythwaites) [inline apps/sse/src/sse_util_os.erl]: Width 80 chars. Verticalise thus:
```
MaxBytes = maps:get(max_size, Opts, infinity),
{Status, Bytes} = get_data(
  Port, MonRef, Eot, ExitStatusOpt, [], 0, MaxBytes),
```

File: apps/sse/src/sse_util_os.erl [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+      Options :: os_command_opts(),
+      Out :: string(),
+      Status :: integer() | undefined.
+cmd(Cmd, Opts) ->
+    ExitStatusOpt =
+        case maps:get(exit_status, Opts, false) of
+            false -> false;
+            _ -> true
+        end,
+    {SpawnCmd, SpawnOpts, SpawnInput, Eot} = mk_cmd(os:type(), ExitStatusOpt, validate(Cmd)),
```
Comment (jacobythwaites) [inline apps/sse/src/sse_util_os.erl]: Verticalise max 80 cols width

File: apps/sse/src/sse_util_os.erl [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+%%   This is because deviations from OTP would be untested on Windows.
+%% - Return exit status when the option "exit_status" is "true".
+%%   When option "exit_status" is "true", the function may hang
+%%   if a background command ("&") is in the given command.
+%%   If the given command closes its standard output without exiting,
+%%   the behaviour of this function is unspecified; if this function
+%%   returns, it will most likely return the atom "undefined" as exit
+%%   status.  This depends on the behaviour of the underlying
+%%   "open_port" function.
+-spec cmd(Command) -> string() when
```
Comment (jacobythwaites) [inline apps/sse/src/sse_util_os.erl]: Verticalise thus:
```
-spec cmd(os_command()) ->
  string().
```

File: apps/sse/src/sse_util_os.erl [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+      Command :: os_command().
+cmd(Cmd) ->
+    cmd(Cmd, #{ }).
```
Comment (jacobythwaites) [inline apps/sse/src/sse_util_os.erl]: Indent 2 spaces not 4

File: apps/sse/src/sse_util_os.erl [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+
+-spec cmd(Command, Options) -> Out | {ok, Status, Out} when
```
Comment (jacobythwaites) [inline apps/sse/src/sse_util_os.erl]: Verticalise thus:
```
-spec cmd(Command, Options) ->
  Out
| {ok, Status, Out}
when
  ...
```
I don't really like return values that aren't consistent with the `{ok, ..} | {error, reason()}` convention. Could we have the status always returned?

File: apps/sse/src/sse_util_os.erl [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+            flush_until_down(Port, MonRef);
+        {'DOWN', MonRef, _, _, _} ->
+            flush_exit(Port)
+    end.
+
+%% The exit signal is always delivered before
+%% the down signal, so we can be sure that if there
+%% was an exit message sent, it will be in the
+%% mailbox now.
+flush_exit(Port) ->
```
Comment (jacobythwaites) [inline apps/sse/src/sse_util_os.erl]: 2 space indent here and elsewhere

File: apps/sse_cfg/src/sse_cfg_transform.erl [hunk 7]
```diff
@@ -103,13 +103,14 @@ xslt_proc(XsltProc, XmlFile, XslFile, XslParams) ->
...
       {xml, XmlFile}]),
 
-  case os:cmd(Cmd) of
-    [] ->
+  case sse_util:run_cmd_with_status(Cmd) of
+    {ok, 0, _} ->
       {ok, OutFile};
 
-    Output ->
+    {ok, Status, Output} ->
```
Comment (jacobythwaites) [inline apps/sse_cfg/src/sse_cfg_transform.erl]: Nice

File: apps/sse/src/sse_util_os.erl (hunk 5 above)
Comment (lucafavatellawork) [reply apps/sse/src/sse_util_os.erl]: > I don't really like return values that aren't consistent with the `{ok, ..} | {error, reason()}` convention.

The idea is to keep the function compatible with the [OTP `os:cmd/2`](https://erlang.org/doc/man/os.html#cmd-2), so to eventually be able to contribute these changes upstream. `os:cmd/2` returns `string()`, hence when this function is called in the same way as `os:cmd/2` - i.e. without `exit_status` option enabled - it returns the same thing as `os:cmd/2` i.e. a `string()`. This is where the strange return type comes from.

>  Could we have the status always returned?

This function is called only by the `sse_util` module, from the `run_cmd_with_status/1` function, and such function returns a consistent type of `{ok, integer() | undefined, string()}`. May I suggest this is enough?

I am preparing a patch for the verticalisation.

File: apps/sse/src/sse_util_os.erl (hunk 1 above)
Comment (lucafavatellawork) [reply apps/sse/src/sse_util_os.erl]: Addressed in new commits I pushed.

File: apps/sse/src/sse_util_os.erl (hunk 2 above)
Comment (lucafavatellawork) [reply apps/sse/src/sse_util_os.erl]: Addressed in new commits I pushed.

File: apps/sse/src/sse_util_os.erl (hunk 3 above)
Comment (lucafavatellawork) [reply apps/sse/src/sse_util_os.erl]: Addressed in new commits I pushed.

File: apps/sse/src/sse_util_os.erl (hunk 4 above)
Comment (lucafavatellawork) [reply apps/sse/src/sse_util_os.erl]: Addressed in new commits I pushed.

File: apps/sse/src/sse_util_os.erl (hunk 5 above)
Comment (lucafavatellawork) [reply apps/sse/src/sse_util_os.erl]: > Verticalise thus:

Addressed in new commits I pushed.

> I don't really like return values that aren't consistent with the {ok, ..} | {error, reason()} convention. Could we have the status always returned?

@jacobythwaites I did not perform amendments for this for now - please refer to [previous comment](https://github.com/opensparkl/sse_core/pull/62#discussion_r359795662). Happy to reconsider though.

File: apps/sse/src/sse_util_os.erl (hunk 6 above)
Comment (lucafavatellawork) [reply apps/sse/src/sse_util_os.erl]: Addressed in new commits I pushed.

File: apps/sse/src/sse_util.erl [hunk 8]
```diff
@@ -708,7 +708,7 @@ run_cmd(string()) ->
 
 run_cmd(Cmd) ->
   Result =
-    os:cmd(Cmd),
+    sse_util_os:cmd(Cmd),
```
[the remark this answers is not in the document]
Comment (lucafavatellawork) [reply apps/sse/src/sse_util.erl]: I see this comment is on an old version of the code so I did not act on it.

The docs and specs of the version of the function in the current state of this PR are:
```
%% @doc
%% Runs the command, returning ok if the exit status is zero.
%%
%% It also generates a log entry with the command, the exit status and
%% the output.
%%
%% If the command contains a background command (`&') or closes its
%% standard output without exiting, the behaviour is unspecified.
-spec
run_cmd(string()) ->
  ok | {error, integer() | undefined}.
```

File: apps/sse/src/sse_util_os.erl (hunk 1 above)
Comment (lucafavatellawork) [reply apps/sse/src/sse_util_os.erl]: Amended in https://github.com/opensparkl/sse_core/pull/62/commits/ff957f2a3f96aa4281c67e163fad406c49d609d2

**Pull request merged**